mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,494 @@
+          PartitionField partitionField = partitionFields.get(j);
+          if (identityColumn.fieldId() == partitionField.sourceId() &&
+              "identity".equals(partitionField.transform().toString())) {
+            row.set(size + i, partition.get(j, spec.javaClasses()[i]));
+          }
+        }
+      }
+      return row;
+    }
+
+    private CloseableIterable<T> newAvroIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      Avro.ReadBuilder avroReadBuilder = Avro.read(inputFile)
+                                             .project(readSchema)
+                                             .split(task.start(), task.length());
+
+      if (reuseContainers) {
+        avroReadBuilder.reuseContainers();
+      }
+
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          //TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          avroReadBuilder.createReaderFunc(DataReader::create);
+      }
+      return avroReadBuilder.build();
+    }
+
+    private CloseableIterable<T> newParquetIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      Parquet.ReadBuilder parquetReadBuilder = Parquet.read(inputFile)
+                                                      .project(readSchema)
+                                                      .filter(task.residual())
+                                                      .caseSensitive(caseSensitive)
+                                                      .split(task.start(), task.length());
+      if (reuseContainers) {
+        parquetReadBuilder.reuseContainers();
+      }
+
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          //TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          parquetReadBuilder.createReaderFunc(
+              fileSchema -> GenericParquetReaders.buildReader(readSchema, fileSchema));
+      }
+      return parquetReadBuilder.build();
+    }
+
+    private CloseableIterable<T> newOrcIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      ORC.ReadBuilder orcReadBuilder = ORC.read(inputFile)
+                                          .schema(readSchema)
+                                          .caseSensitive(caseSensitive)
+                                          .split(task.start(), task.length());
+      // ORC does not support reuse containers yet
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          //TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          //TODO: We do not have support for Iceberg generics for ORC
+          throw new UnsupportedOperationException();
+      }
+
+      return orcReadBuilder.build();
+    }
+  }
+
+  private static Table getTable(Configuration conf) {
```

Comment:
In Spark, this is an instance method so that user can override it and use their own catalog. I think that would be a good idea here as well:

```java
public class CustomCatalogIcebergInputFormat extends IcebergInputFormat {
  @Override
  public Table getTable(String tablePath) {
    CustomCatalog.get().loadTable(conf.get(tablePath));
  }
}
```

In that example, I'm assuming that we'd have a `final getTable(Configuration)` method that pulls `TABLE_PATH` out of the config and passes it to `getTable`.

Reply:
I think this is slightly related to the HiveCatalog discussion above where we want to load HiveCatalog through reflection if HiveCatalog is on the classpath [since we do not want to add the iceberg-hive dependency]

The problem here is that this `getTable` method needs to be static since it is being called from the `configBuilder`.

To solve both these issues - not relying on hive-deps at compile time and ability to plug in a custom catalog, can we define a `CatalogFactory` which allows us to load our own custom catalogs. The `CatalogFactory` can be passed as a config property. Since this factory is an interface, it should be easier to instantiate [through reflection] and then we can load our custom catalog.

What do u think?

Reply:
Yeah, I guess that's not a good way to inject a catalog. In addition, the configuration methods sets up the `IcebergInputFormat` and not a user's subclass.

I think the factory idea would work, but it's more of a `Supplier<Catalog>` since there is no configuration. Like this?

```java
  ConfigBuilder catalogSupplier(Class<? extends Supplier<Catalog>> supplierClass)) {
    this.supplierClass = supplierClass;
    return this;
  }
```

Reply:
Supplier can work, but how do we pass the Configuration to the  Catalog  to instantiate. Supplier is a function with no args. Maybe a `Function<Configuration,Catalog>` works better?

Reply:
You're right, we probably want to use `Function` instead.